### ISSUE TITLE
[Bug] Console not usable after ALT-Enter in Win32 [sf#184]

### ISSUE THREAD

**Issue by openMSX-import** [openMSX/openMSX#485]
**Reported by manuelbi on 2005-03-05 18:12 UTC**
After pressing ALt-Enter for full screen, the Enter key
seems to hang, which makes the console unusable (it
doesn't come up).
SDL bug?

**Comment by openMSX-import** [comment]
**Commented by mfeingol on 2010-02-15 21:09 UTC**
I think this is an SDL bug.  I can't reproduce the precise behavior you mention on Win7 - for me the ALT key gets stuck.

And I think we know at this point that the ALT key getting stuck (at least on Windows) is an SDL bug.  Sometime when http://www.braingames.getput.com/mog/ is up, I'll go look for my original bug report and see if there's anything further we can do.

**Comment by openMSX-import** [comment]
**Updated by mfeingol on 2010-02-15 21:13 UTC**
- **assigned_to**: nobody --> mfeingol

**Comment by openMSX-import** [comment]
**Commented by mfeingol on 2010-02-15 21:43 UTC**
For posterity, this is the relevant thread:  http://www2.braingames.getput.com/forum/forum\_posts.asp?TID=114&PN=4

**Comment by openMSX-import** [comment]
**Commented by mfeingol on 2010-02-15 23:39 UTC**
When openmsx is full screen and alt+enter is pressed, SDL detects that the ALT key has been released, but can't enqueue the event because its event queue is still inactive.  This is because transitioning from full screen to windowed mode deactivates the event queue while the video subsystem is being destroyed, then recreates it.

This appears to be timing dependent.  On my system, the ALT released event is processed when going full screen, but dropped when returning to windowed mode.  By introducing breakpoints, I can cause the ALT released event to be dropped when going full screen as well.  I've even seen cases where from SDL's perspective, the ALT key is simply stuck.  Again, all timing related.

I'm not sure what to suggest here...  A possible workaround might be to simulate ALT release presses when switching to full screen mode and back.  It should be possible to detect that openmsx or SDL itself is out of sync with the OS in these cases (e.g. by calling GetKeyState directly on Win32) and take the appropriate corrective action.

The bottom line is that detecting key states based on window events is a bad idea on Windows because input queues are tied to windows and windows can be created and destroyed, leaving timing windows.  Of course, that's precisely what SDL does when going full screen.

I have to conclude that full screen will likely never work 100% reliably with SDL 1.2 on Windows.  Other systems may vary.

**Comment by openMSX-import** [comment]
**Updated by mfeingol on 2010-02-15 23:39 UTC**
- **priority**: 2 --> 5

**Comment by MBilderbeek** [comment]
User "Creepy" couldn't reproduce this anymore with a recent version on Windows 11. So this seems to be fixed. If someone can still reproduce this, please reopen.
